src/DeviceInput/deviceInputSystem.ts

```diff
@@ -0,0 +1,225 @@
+import { Engine } from '../Engines/engine';
+import { Nullable } from '../types';
+
+/**
+ * This class will take all inputs from Keyboard, Pointer, and
+ * any Gamepads and provide a polling system that all devices
+ * will use.  This class assumes that there will only be one
+ * pointer device and one keyboard.
+ */
+export class DeviceInputSystem {
+    // Private Members
+    private _inputs: Map<string, Array<number>> = new Map();

```

Comment:
Just out of curiosity, what are our minimum browser support requirements?

Reply:
I've been assuming IE 11+, Firefox 4+, Google Chrome 9+ and Opera 15+

Reply:
Nope you are right

Reply:
> Is there something Map gives you that an {} would not provide?

Maps are better for a number of reasons, as noted in the "Objects vs. Maps" [here](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Map).

The issue is whether is it well-supported in browsers that we support. If you look at the "Browser compatibility" section at the end of that link, you'll see that IE11 has only partial support, which may cause weird issues that are hard to spot.

Reply:
Ok, I see that Maps are preferable in certain situations, but I don't see the special properties being used here? There isn't a reference to its size, I don't see any non string or symbol keys, I don't think there are going to be a ton of removals/additions of key-value pairs.

Happy to be be wrong, just trying to learn why why Map was used over {}

Reply:
Thanks for the info!  Maps won't work so I'll get that changed over to an associative array.

Reply:
> Maps are preferable in certain situations

If we didn't have to worry about IE, then we probably would be using Maps for, well, maps instead of using Objects.

Reply:
I think my over-familiarity with JS is coming back to bite me. In my head, maps and objects were pretty much interchangeable terms for the same thing (when used like this). I did not realize there was such a distinction. TIL.